cmd/kubeadm/app/phases/kubelet/kubelet.go

```diff
@@ -86,6 +90,15 @@ func UpdateNodeWithConfigMap(client clientset.Interface, nodeName string) error
 			return false, nil
 		}
 
+		baseCongfig, err := yaml.Marshal(kubeletCfg.Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey])
```

Comment:
I think you misunderstood this. The configuration that should be written to the file is `cfg.KubeletConfiguration.BaseConfig`; basically write the same thing you have in the configmap on disk in the init code.

In the join code, this config should be looked up from the configmap (we need a new RBAC rule for that)

Reply:
The `kubeletCfg` here is the configmap we stored in API server, it's `Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey]` is the BaseConfig which we want.

See Line 50:
```go
kubeletBytes, err := kubeadmutil.MarshalToYamlForCodecs(cfg.KubeletConfiguration.BaseConfig, kubeletconfigv1alpha1.SchemeGroupVersion, *kubeletCodecs)
```
and Line 60-61:
```go
 		Data: map[string]string{
 			kubeadmconstants.KubeletBaseConfigurationConfigMapKey: string(kubeletBytes),
 		},
```

Feel free to correct me if I'm indeed misunderstood here :)

Reply:
>we need a new RBAC rule for that

What RBAC rule is needed? The rules for group "system:nodes" have already been created in init (`CreateBaseKubeletConfiguration`). Am I missing sth here?
```go
 	if err := createKubeletBaseConfigMapRBACRules(client); err != nil {
 		return fmt.Errorf("error creating base kubelet configmap RBAC rules: %v", err)
 	}
```

Reply:
wrt RBAC, see the instructions regarding permissions here: https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

Reply:
Does that mean we should create separate rules for each node?

Reply:
Ok, I'll clarify this and alter my earlier plan a bit to make it better.
Let me know if this sounds good to you

1) There should be two functions: one that takes a `string` to write to disk (`WriteInitKubeletConfigToDisk`) and one for `kubeadm join` that takes a k8s client which extracts the config string (`GetKubeletConfigFromConfigMap`)
2) `WriteInitKubeletConfigToDisk` should:
 - Write the string down to this file `filepath.Join(KubeletBaseConfigurationDir, KubeletBaseConfigurationFile)`
3) `GetKubeletConfigFromConfigMap` should:
 - Get the base-config ConfigMap from the API server using a client that is passed in
 - Extract the value string from the `kubeadmconstants.KubeletBaseConfigurationConfigMapKey`

No `yaml.Marshal` or similar code should be used here.